FEAT-#5759: Implement lazy Arrow execution for the HDK engine

AndreyPavlenko opened the pull request:
- Reimplemented _union_all(), that used non-lazy arrow-based frames concatenation.
 - Added functionality to dataframe tree nodes and expressions for arrow execution. The arrow execution is performed if the frame has an arrow table in partitions and all the nodes support the arrow execution. It allows to save on the table import to HDK.
 - The arrow execution is also performed for the operations, not supported by HDK, for example, bitwise not '~'.
 
## What do these changes do?
#### [dataframe.py](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0):
[_execute()](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0R1842). The main changes here:
- Now the method returns either `pyarrow.Table` or `pandas.Dataframe`. Dataframe is returned if we have an unsupported data.
- Added new condition - `if not frame._op.can_execute_hdk()`. If it's true, then the operation can not be executed with HDK and should be executed with arrow. Currently, there is only one such operation - `~` `invert()`. The frame is added to the stack for the subsequent materialization.

[_materialize()](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0R1871):
- The method returns `pyarrow.Table`
- Added new execution mode - `hdk`. This is used by unit tests to force the HDK execution.

[_require_executed_base(), _can_execute_arrow(), execute_arrow()](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0L2095) moved to [DFAlgNode](https://github.com/modin-project/modin/pull/6251/files#diff-fc0f76c51d86e975ebb2af41cc109d92311368a9da5fc00a4b264f85d8bce737R239) and subclasses.

[_arrow_row_slice()](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0L2264) moved to [MaskNode.execute_arrow()](https://github.com/modin-project/modin/pull/6251/files#diff-fc0f76c51d86e975ebb2af41cc109d92311368a9da5fc00a4b264f85d8bce737R539)

[_arrow_select()](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0L2225) moved to [TransformNode.execute_arrow()](https://github.com/modin-project/modin/pull/6251/files#diff-fc0f76c51d86e975ebb2af41cc109d92311368a9da5fc00a4b264f85d8bce737R715) . The new implementation is much more simple, because the actual logic is now implemented in subclasses of [BaseExpr.execute_arrow()](https://github.com/modin-project/modin/pull/6251/files#diff-532f2950a0930fecba54c8ab7a130cd9baa85d9320e8b46d9035931be52bc987R668). Each implementation calculates an arrow column and this method just concatenates all the columns and creates a new arrow table.

[_union_all_arrow()](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0L1195) moved to [UnionNode.execute_arrow()](https://github.com/modin-project/modin/pull/6251/files#diff-fc0f76c51d86e975ebb2af41cc109d92311368a9da5fc00a4b264f85d8bce737R937). The new implementation is also much more simple, because it receives already computed tables. The actual tables computation is performed by the corresponding `execute_arrow()` implementation for each frame.

#### [df_algebra.py](https://github.com/modin-project/modin/pull/6251/files#diff-fc0f76c51d86e975ebb2af41cc109d92311368a9da5fc00a4b264f85d8bce737):
Added 4 new methods to `DFAlgNode` and subclasses. When processing the execution tree and all the nodes `can_execute_arrow()`, the execution is performed with `execute_arrow()`, but not HDK. Depending on the number of input nodes, this method receives either a single or multiple arrow tables, then performs the required computation and returns a single table.

#### [expr.py](https://github.com/modin-project/modin/pull/6251/files#diff-532f2950a0930fecba54c8ab7a130cd9baa85d9320e8b46d9035931be52bc987):
- Similarly to `DFAlgNode`, new arrow methods added to `BaseExpr`. Since `BaseExpr` represents a single column, the `execute_arow()` method returns a ChunkedArray.
- Added new method `fold()`. It allows, in some cases, to replace multiple expressions with a single one or with a precomputed value. For example, if we have multiple subsequent arithmetic expressions with literals, we can precompute the result and replace all the expressions with a single literal.

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` <!-- you can amend your commit with a signature via `git commit -amend -s` -->
- [x] Resolves #5759 <!-- issue must be created for each patch -->
- [x] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date <!-- if you have added, renamed or removed files or directories please update the documentation accordingly -->

Garra1980 commented on modin/pandas/test/dataframe/test_reduce.py:
```diff
@@ -98,6 +98,10 @@ def test_count_specific(numeric_only):
 
+@pytest.mark.skipif(
+    StorageFormat.get() == "Hdk",
+    reason="https://github.com/intel-ai/hdk/issues/513",
```
this one is fixed

anmyachev commented:
@AndreyPavlenko this pull request is very large, how many logical parts can you split it into?

anmyachev commented:
> > @AndreyPavlenko this pull request is very large, how many logical parts can you split it into?
> 
> There are several logical parts in - modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:
> 
> * New implementation of _union_all(). This is the biggest part in this file, but the new implementation is much smaller than the previous one.
> * Changes to _execute() and _materialize(), related to the arrow-based execution logic.
> * Added new methods - isna() and invert(). This part is quite simple.
> 
> Although there are a lot of changes in this file, most of the changes are deletions. The code, related to the arrow execution, moved to df_algebra.py.
> 
> New methods added to df_algebra.py and expr.py. These methods contains arrow-based execution (partially, the code moved from dataframe.py) and expressions folding.

Why don't we split this pull request into 3 parts in that case?

AndreyPavlenko commented:
> Why don't we split this pull request into 3 parts in that case?

They depend on each other. A lot of code from dataframe.py moved to df_algebra.py and expr.py. I could move isna() and invert() into a separate PR, but does it make sense? The code is quite simple. Moreover, these methods are used for testing of the main functionality, introduced by this PR.

anmyachev commented:
> Although there are a lot of changes in this file, most of the changes are deletions. The code, related to the arrow execution, moved to df_algebra.py.

Moving the code can be put in a separate pull request, right? This will greatly facilitate the review, since there is no need to check the code that has not changed, but is highlighted here.

AndreyPavlenko commented:
> Moving the code can be put in a separate pull request, right?

The code is not just moved but also changed. For example:

[_arrow_row_slice](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0L2264) moved to [execute_arrow](https://github.com/modin-project/modin/pull/6251/files#diff-fc0f76c51d86e975ebb2af41cc109d92311368a9da5fc00a4b264f85d8bce737R539)

[_arrow_select](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0L2225) moved to [execute_arrow](https://github.com/modin-project/modin/pull/6251/files#diff-fc0f76c51d86e975ebb2af41cc109d92311368a9da5fc00a4b264f85d8bce737R715)

[_union_all_arrow](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0L1195) moved to [execute_arrow](https://github.com/modin-project/modin/pull/6251/files#diff-fc0f76c51d86e975ebb2af41cc109d92311368a9da5fc00a4b264f85d8bce737R937)

The last 2 have been changed significantly.

anmyachev commented:
> [_arrow_row_slice](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0L2264) moved to [execute_arrow](https://github.com/modin-project/modin/pull/6251/files#diff-fc0f76c51d86e975ebb2af41cc109d92311368a9da5fc00a4b264f85d8bce737R539)

Let's start simplifying this pull request with `_arrow_row_slice`. Its relocation as well as the introduction of a new interface for all classes inherited from `DFAlgNode`.

AndreyPavlenko commented:
> Let's start simplifying this pull request with `_arrow_row_slice`. Its relocation as well as the introduction of a new interface for all classes inherited from `DFAlgNode`.

Do you think it could simplify review? _arrow_row_slice is almost not changed but just moved to a different file with minimal changes.

anmyachev commented:
> > Let's start simplifying this pull request with `_arrow_row_slice`. Its relocation as well as the introduction of a new interface for all classes inherited from `DFAlgNode`.
> 
> Do you think it could simplify review? _arrow_row_slice is almost not changed but just moved to a different file with minimal changes.

Yes, however, this is not the last part that I will ask to be taken out in a separate pull request.

Please also keep in mind that several simple pull requests are easier to accept and merge than one large.

AndreyPavlenko commented:
> Please also keep in mind that several simple pull requests are easier to accept and merge than one large.

I understand that, but, from the other side, the PR should contain a completed feature. This PR reimplements the arrow-based execution logic.  Obviously, it affects the old implementation of the arrow execution. What can be moved out of this PR are isna() and invert(), but I believe it will not simplify the code review at all. Moreover, these methods are used for testing of the new functionality.

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/base_worker.py:
```diff
@@ -251,4 +251,4 @@ def import_pandas_dataframe(cls, df, name=None):
             Imported table name.
         """
-        return cls.import_arrow_table(pa.Table.from_pandas(df))
+        return cls.import_arrow_table(pa.Table.from_pandas(df), name=name)
```
Why was this option not used before?

AndreyPavlenko replied:
Seems it was missing.

anmyachev replied:
Why are you making this change now? Do you have a test that doesn't work without it?

anmyachev replied:
In this case, again logically unrelated things :(

AndreyPavlenko replied:
Do you think a separate PR is required for such a minor and obvious fix?

AndreyPavlenko replied:
I agree, but this is an internal code and should not affect users.

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/test/utils.py:
```diff
@@ -272,16 +274,21 @@ def run_modin(
         kwargs["df2"] = pd.DataFrame(data2, **constructor_kwargs)
         kwargs["df"] = kwargs["df1"]
 
-        if force_lazy:
```
By moving `force_lazy` code branch after `force_arrow_execute` code branch, you implicitly change the execution of some tests. For example, `test_empty_transform` and `test_append`.

To avoid this, you can make a check that only one of the 3 parameters (`force_lazy`, `force_hdk_execute`, `force_arrow_execute`) can be true.

AndreyPavlenko replied:
Yes and it seems correct. Previously, `force_arrow_execute` was always ignored, unless `force_lazy` was set to `False` explicitly, because it's `True` by default. The mentioned tests set `force_arrow_execute=True`, but do not set `force_lazy=False`. Thus, the `force_arrow_execute` parameter was simply ignored.

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/calcite_serializer.py:
```diff
@@ -189,7 +189,10 @@ def serialize_obj(self, obj):
         for k, v in obj.__dict__.items():
             if k[0] != "_":
-                res[k] = self.serialize_item(v)
+                if k == "op" and isinstance(obj, OpExpr) and v == "//":
+                    res[k] = "/"
```
Why serialization in this case is done separately and not in `serialize_item` function?

AndreyPavlenko replied:
serialize_item receives a string. We can not replace any string "//" with "/". This is only done for a single case - the operation name of OpExpr.

anmyachev replied:
Why did this new case appear?

AndreyPavlenko replied:
Added support for floordiv [here](https://github.com/modin-project/modin/pull/6251/files#diff-532f2950a0930fecba54c8ab7a130cd9baa85d9320e8b46d9035931be52bc987R165) and [here](https://github.com/modin-project/modin/pull/6251/files#diff-532f2950a0930fecba54c8ab7a130cd9baa85d9320e8b46d9035931be52bc987R923)

anmyachev replied:
Now I see, however it looks like a workaround to the fact that coded operations have appeared that cannot be processed in `serialize_item` function.

AndreyPavlenko replied:
Right, it's a workaround.

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/utils.py:
```diff
@@ -20,22 +20,25 @@
 
-import numpy as np
-
 import pandas
```
```suggestion
import numpy as np
import pandas
```

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/utils.py:
```diff
@@ -20,22 +20,25 @@
 import pyarrow as pa
 from pyarrow.types import is_dictionary
 
+import numpy as np
+
```
```suggestion
```

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/utils.py:
```diff
@@ -20,22 +20,25 @@
+
 from modin.utils import MODIN_UNNAMED_SERIES_LABEL
 
+_EMPTY_ARROW_TABLE = None
```
Why not? Instead of using `empty_arrow_table`.
```suggestion
_EMPTY_ARROW_TABLE = pa.Table.from_pandas(pandas.DataFrame({}))
```

AndreyPavlenko replied:
Just to save memory a little bit. This empty table may never be used at runtime or used quite rarely, thus, it's created lazy, on the first access.

anmyachev replied:
What is the amount of memory you want to save? If it is not significant, then let's not complicate the code.

AndreyPavlenko replied:
No, it's not significant.

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:
```diff
@@ -82,12 +78,11 @@
+UNNAMED_IDX_COL_NAME = ColNameCodec.UNNAMED_IDX_COL_NAME
 encode_col_name = ColNameCodec.encode
 decode_col_name = ColNameCodec.decode
-concat_index_names = ColNameCodec.concat_index_names
```
Why?

AndreyPavlenko replied:
It's used only in a single method _union_all().

AndreyPavlenko replied:
Right, but it would be reasonable to have this alias in the function scope. Or remove it at all.

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:
```diff
@@ -1038,350 +1071,107 @@ def _union_all(
         """
-        frames = [self] + other_modin_frames
-
-        # This is a special case, where we need to preserve the index of empty frames.
```
How is this case handled now?

AndreyPavlenko replied:
[Here](https://github.com/modin-project/modin/pull/6251/files#diff-fc0f76c51d86e975ebb2af41cc109d92311368a9da5fc00a4b264f85d8bce737R950)

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:
```diff
@@ -1038,350 +1071,107 @@ def _union_all(
+        index_cols = None
+        columns = OrderedDict()
+        for col in self.columns:
+            columns[col] = self._dtypes[col]
```
Why is the variable named like `columns`, but types are stored inside?

anmyachev replied:
Non-obvious variable name. Maybe `columns2types`?

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:
```diff
@@ -1038,350 +1071,107 @@ def _union_all(
+            raise NotImplementedError(f"Unsupported join type {join=}")
+
+        frames = []
+        for frame in [self] + other_modin_frames:
+            # Empty frames are filtered out only in case of the outer join.
```
Has there been filtering for empty frames in this function before, and if so, where?

AndreyPavlenko replied:
Here - [_union_all_arrow()](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0L1256)

AndreyPavlenko commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:
```diff
@@ -238,7 +233,7 @@ def __init__(
             self.set_dtypes_cache(dtypes)
 
-        if partitions is not None:
+        if partitions is not None and partitions.size > 1:
```
It doesn't make sense to call `_filter_empties()` in case of a single partition. Moreover, we should never have more than one partition in the HDK backend, that is not true at the moment and will be fixed in a separate PR. This code will be remover either.

anmyachev replied:
ok

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:
```diff
@@ -1038,350 +1071,107 @@ def _union_all(
+                index_cols = list(index_names)
+                dtypes = pd.Series(index_names.values(), index=index_cols)
+        else:
+            # Find common dtypes
```
Why is this section called `Find common dtypes`, but only `columns` variable is changed?

AndreyPavlenko replied:
`columns` is a dictionary with the column names and dtypes.

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:
```diff
@@ -1038,350 +1071,107 @@ def _union_all(
-                if idx_equal:
-                    idx_cols = list(
-                        concat_index_names([f.frame for f in frames]).keys()
+                table_columns = ColNameCodec.concat_index_names(frames)
+                index_cols = list(table_columns)
```
This looks clearer
```suggestion
                index_cols = ColNameCodec.concat_index_names(frames)
                table_columns = list(index_cols)
```

AndreyPavlenko replied:
Sorry, but it's not correct. `table_columns` should be a dictionary, and `index_cols` - a list.

anmyachev replied:
`table_columns2types`?

AndreyPavlenko commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:
```diff
@@ -1038,350 +1071,107 @@ def _union_all(
-
-            return new_columns, new_dtypes
-
-        # If all frames are either FrameNode(materialized frame) or UnionNode,
-        # containing only FrameNodes and having the same concatenation options,
-        # put all frames into a single UnionNode. It allows to concatenate
-        # multiple frames with arrow in a single batch operation.
-        materialized = []
```
This logic is now implemented globally. If all nodes are materialized and could be computed with arrow, the arrow execution is used. The `UnionNode.execute_arrow()` receives a list of arrow tables for each frame.

anmyachev replied:
One point of code through which the flow of execution passes?

AndreyPavlenko replied:
Global logic for all nodes. Now each DFAlgNode have a new method - `can_execute_arrow`. When traversing the execution tree, we check this condition and if it's true for each node, then the arrow execution is performed.

AndreyPavlenko commented on modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:
```diff
@@ -1038,350 +1071,107 @@ def _union_all(
-            )
+                columns = OrderedDict((col, columns[col]) for col in sorted(columns))
 
-        # In case of different number of columns, HDK performs
-        # slowly and supports only numeric column types.
-        # See https://github.com/intel-ai/hdk/issues/182
-        # To work around this issue, perform concatenation
-        # with arrow.
```
Here - [UnionNode.can_execute_hdk()](https://github.com/modin-project/modin/pull/6251/files#diff-fc0f76c51d86e975ebb2af41cc109d92311368a9da5fc00a4b264f85d8bce737R915)

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py:
```diff
@@ -84,10 +84,11 @@ def executeDML(cls, query):
     def executeRA(cls, query):
         if query.startswith("execute relalg"):
-            ra = query.removeprefix("execute relalg")
+            ra = query[14:]
```
Why?

AndreyPavlenko replied:
This code failed on Python 3.8, because `removeprefix()` added since 3.9.

anmyachev commented on modin/experimental/core/execution/native/implementations/hdk_on_native/hdk_worker.py:
```diff
@@ -84,10 +84,11 @@ def executeDML(cls, query):
     def executeRA(cls, query):
         if query.startswith("execute relalg"):
-            ra = query.removeprefix("execute relalg")
+            ra = query[14:]
```
```suggestion
            # 14 == len("execute relalg") + 1
            ra = query[14:]
```

AndreyPavlenko commented on modin/experimental/core/execution/native/implementations/hdk_on_native/df_algebra.py:
```diff
@@ -15,10 +15,25 @@
 
-from .dataframe.utils import ColNameCodec
-from .expr import InputRefExpr
+import typing
+from typing import TYPE_CHECKING, List, Dict, Union
```
It conflicts with the next import:
from collections import OrderedDict
